// @bit/mui-org.material-ui.styles/dist/styles/createMixins.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createMixins;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function createMixins(breakpoints, spacing, mixins) {
  var _toolbar;

  return Object.assign({
    gutters: function gutters() {
      var styles = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      return Object.assign({
        paddingLeft: spacing.unit * 2,
        paddingRight: spacing.unit * 2
      }, styles, _defineProperty({}, breakpoints.up('sm'), Object.assign({
        paddingLeft: spacing.unit * 3,
        paddingRight: spacing.unit * 3
      }, styles[breakpoints.up('sm')])));
    },
    toolbar: (_toolbar = {
      minHeight: 56
    }, _defineProperty(_toolbar, breakpoints.up('xs') + ' and (orientation: landscape)', {
      minHeight: 48
    }), _defineProperty(_toolbar, breakpoints.up('sm'), {
      minHeight: 64
    }), _toolbar)
  }, mixins);
}
